In [51]:
import pandas as pd
import numpy as np

In [52]:
chipotle = pd.read_csv('chipotle.tsv', sep='\t')
chipotle

order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
1            1         1                                   Izze   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
4            2         2                           Chicken Bowl   
...        ...       ...                                    ...   
4617      1833         1                          Steak Burrito   
4618      1833         1                          Steak Burrito   
4619      1834         1                     Chicken Salad Bowl   
4620      1834         1                     Chicken Salad Bowl   
4621      1834         1                     Chicken Salad Bowl   

                                     choice_description item_price  
0                                                   NaN     $2.39   
1                                          [Clementine]     $3.39   
2                                               [Apple]     $3.39   
3                                                   NaN     $2.39   
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   
...                                                 ...        ...  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...    $11.75   
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...    $11.75   
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25   
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75   
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75   

[4622 rows x 5 columns]

## 1. Missing Values:


In [53]:
chipotle[chipotle.isna()]

order_id  quantity item_name choice_description item_price
0          NaN       NaN       NaN                NaN        NaN
1          NaN       NaN       NaN                NaN        NaN
2          NaN       NaN       NaN                NaN        NaN
3          NaN       NaN       NaN                NaN        NaN
4          NaN       NaN       NaN                NaN        NaN
...        ...       ...       ...                ...        ...
4617       NaN       NaN       NaN                NaN        NaN
4618       NaN       NaN       NaN                NaN        NaN
4619       NaN       NaN       NaN                NaN        NaN
4620       NaN       NaN       NaN                NaN        NaN
4621       NaN       NaN       NaN                NaN        NaN

[4622 rows x 5 columns]

In [54]:
chipotle.loc[chipotle.choice_description.isna(), 'choice_description']

0       NaN
3       NaN
6       NaN
10      NaN
14      NaN
       ... 
4600    NaN
4605    NaN
4613    NaN
4614    NaN
4616    NaN
Name: choice_description, Length: 1246, dtype: object

We are not sure what to impute in missing values so we will remove the misssing values in `choice_description` column

In [55]:
chipotle.dropna(subset='choice_description', inplace=True)

## 2. Data Types:

In [56]:
chipotle.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3376 entries, 1 to 4621
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   order_id            3376 non-null   int64 
 1   quantity            3376 non-null   int64 
 2   item_name           3376 non-null   object
 3   choice_description  3376 non-null   object
 4   item_price          3376 non-null   object
dtypes: int64(2), object(3)
memory usage: 158.2+ KB


Converting `order_id` col. to string

In [57]:
chipotle.order_id = chipotle.order_id.astype('str')

## 3. Duplicated Entries:

In [58]:
chipotle.duplicated().value_counts()

False    3335
True       41
Name: count, dtype: int64

We have 41 no. o duplicated entries

In [59]:
chipotle[chipotle.duplicated()]

order_id  quantity            item_name  \
238       103         1        Steak Burrito   
248       108         1          Canned Soda   
297       129         1        Steak Burrito   
381       165         1    Canned Soft Drink   
484       205         1         Chicken Bowl   
567       233         1    Canned Soft Drink   
684       282         1    Canned Soft Drink   
773       319         1         Chicken Bowl   
908       376         1        Steak Burrito   
924       381         1   Chicken Soft Tacos   
1117      462         1      Chicken Burrito   
1203      491         1      Chicken Burrito   
1245      507         1    Canned Soft Drink   
1381      561         1    Canned Soft Drink   
1420      576         1        Barbacoa Bowl   
1421      576         1  Barbacoa Salad Bowl   
1502      610         1      Chicken Burrito   
1612      651         1                 Izze   
1791      724         1    Canned Soft Drink   
1792      724         1    Canned Soft Drink   
2078      838         1   Chicken Soft Tacos   
2127      856         1      Chicken Burrito   
2307      926         1         Chicken Bowl   
2318      926         1         Chicken Bowl   
2362      940         1    Canned Soft Drink   
2448      972         1         Chicken Bowl   
2553     1011         1    Canned Soft Drink   
2554     1011         1    Canned Soft Drink   
2662     1057         1   Chicken Soft Tacos   
2717     1078         1         Chicken Bowl   
2833     1125         1      Chicken Burrito   
2973     1182         1         Chicken Bowl   
2977     1184         1           Steak Bowl   
3637     1454         1   Chicken Soft Tacos   
3716     1484         1          Canned Soda   
3764     1507         1      Chicken Burrito   
3829     1534         1      Chicken Burrito   
4112     1647         1         Chicken Bowl   
4236     1691         1         Chicken Bowl   
4463     1778         1    Canned Soft Drink   
4603     1827         1     Barbacoa Burrito   

                                     choice_description item_price  
238   [Tomatillo Red Chili Salsa, [Rice, Black Beans...    $11.75   
248                                      [Mountain Dew]     $1.09   
297   [Tomatillo Green Chili Salsa, [Rice, Cheese, G...    $11.75   
381                                              [Coke]     $1.25   
484   [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...     $8.75   
567                                         [Diet Coke]     $1.25   
684                                              [Coke]     $1.25   
773   [Tomatillo-Green Chili Salsa (Medium), Black B...     $8.49   
908   [Roasted Chili Corn Salsa (Medium), [Rice, Faj...     $8.99   
924           [Tomatillo-Red Chili Salsa (Hot), Cheese]     $8.49   
1117  [Fresh Tomato Salsa, [Rice, Black Beans, Chees...    $11.25   
1203  [Fresh Tomato Salsa, [Cheese, Sour Cream, Lett...     $8.75   
1245                                         [Lemonade]     $1.25   
1381                                             [Coke]     $1.25   
1420                         [Roasted Chili Corn Salsa]     $9.25   
1421                         [Roasted Chili Corn Salsa]     $9.39   
1502  [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...     $8.75   
1612                                       [Blackberry]     $3.39   
1791                                             [Coke]     $1.25   
1792                                             [Coke]     $1.25   
2078  [Fresh Tomato Salsa, [Cheese, Sour Cream, Lett...     $8.75   
2127  [Fresh Tomato Salsa, [Rice, Black Beans, Chees...    $11.25   
2307  [Roasted Chili Corn Salsa, [Fajita Vegetables,...     $8.75   
2318  [Roasted Chili Corn Salsa, [Fajita Vegetables,...     $8.75   
2362                                             [Coke]     $1.25   
2448  [Tomatillo Red Chili Salsa, [Rice, Black Beans...     $8.75   
2553                                             [Coke]     $1.25   
2554                                             [Coke] 

Dropping duplicated rows

In [60]:
chipotle.drop_duplicates(inplace=True)

## 4. Quantity and Item Price:

In [76]:
chipotle.quantity.unique()

array([1, 2, 3, 4], dtype=int64)

In [87]:
chipotle.item_price.unique()

array(['$3.39 ', '$16.98 ', '$10.98 ', '$11.75 ', '$9.25 ', '$8.75 ',
       '$11.25 ', '$8.49 ', '$2.18 ', '$8.99 ', '$1.09 ', '$22.50 ',
       '$11.48 ', '$17.98 ', '$17.50 ', '$1.25 ', '$23.78 ', '$6.49 ',
       '$11.08 ', '$22.16 ', '$32.94 ', '$22.20 ', '$10.58 ', '$2.50 ',
       '$23.50 ', '$7.40 ', '$18.50 ', '$6.78 ', '$11.89 ', '$9.39 ',
       '$3.75 ', '$8.69 ', '$8.19 ', '$35.00 ', '$27.75 ', '$26.25 ',
       '$21.96 ', '$4.36 ', '$22.96 ', '$26.07 ', '$12.98 ', '$35.25 ',
       '$33.75 ', '$16.38 ', '$5.00 ', '$8.50 ', '$11.49 '], dtype=object)

## 5. Choice Description:

In [88]:
chipotle[['choice_description']]

choice_description
1                                          [Clementine]
2                                               [Apple]
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...
5     [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...
7     [Tomatillo Red Chili Salsa, [Fajita Vegetables...
...                                                 ...
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...

[3335 rows x 1 columns]

In [90]:
chipotle.choice_description.nunique()

1043

## 6. Handling Special Characters:

In [91]:
chipotle.item_name.unique()

array(['Izze', 'Nantucket Nectar', 'Chicken Bowl', 'Steak Burrito',
       'Steak Soft Tacos', 'Chicken Crispy Tacos', 'Chicken Soft Tacos',
       'Chicken Burrito', 'Canned Soda', 'Barbacoa Burrito',
       'Carnitas Burrito', 'Carnitas Bowl', 'Barbacoa Bowl',
       'Chicken Salad Bowl', 'Steak Bowl', 'Barbacoa Soft Tacos',
       'Veggie Burrito', 'Veggie Bowl', 'Steak Crispy Tacos',
       'Barbacoa Crispy Tacos', 'Veggie Salad Bowl',
       'Carnitas Soft Tacos', 'Chicken Salad', 'Canned Soft Drink',
       'Steak Salad Bowl', '6 Pack Soft Drink', 'Bowl', 'Burrito',
       'Crispy Tacos', 'Carnitas Crispy Tacos', 'Steak Salad',
       'Veggie Soft Tacos', 'Carnitas Salad Bowl', 'Barbacoa Salad Bowl',
       'Salad', 'Veggie Crispy Tacos', 'Veggie Salad', 'Carnitas Salad'],
      dtype=object)

Replacing square brackets with empty string

In [161]:
chipotle.choice_description = chipotle.choice_description.str.replace('[','')
chipotle.choice_description = chipotle.choice_description.str.replace(']','')
chipotle.choice_description.unique()

array(['Clementine', 'Apple',
       'Tomatillo-Red Chili Salsa (Hot), Black Beans, Rice, Cheese, Sour Cream',
       ...,
       'Roasted Chili Corn Salsa, Pinto Beans, Sour Cream, Cheese, Lettuce, Guacamole',
       'Tomatillo Green Chili Salsa, Rice, Black Beans',
       'Tomatillo Green Chili Salsa, Rice, Fajita Vegetables, Black Beans, Guacamole'],
      dtype=object)

Converting `item_price` to a float data type and it represents price in dollars

In [177]:
chipotle.item_price = chipotle.item_price.str[1:].astype('float')

## 7. Order Id Integrity:

Everything looks fine with `order_id`

In [165]:
chipotle[['order_id']]

order_id
1           1
2           1
4           2
5           3
7           4
...       ...
4617     1833
4618     1833
4619     1834
4620     1834
4621     1834

[3335 rows x 1 columns]

## 8. Item Name Standardization:

In [172]:
pd.Series(chipotle.item_name.unique()).sort_values()

25        6 Pack Soft Drink
12            Barbacoa Bowl
9          Barbacoa Burrito
19    Barbacoa Crispy Tacos
33      Barbacoa Salad Bowl
15      Barbacoa Soft Tacos
26                     Bowl
27                  Burrito
8               Canned Soda
23        Canned Soft Drink
11            Carnitas Bowl
10         Carnitas Burrito
29    Carnitas Crispy Tacos
37           Carnitas Salad
32      Carnitas Salad Bowl
21      Carnitas Soft Tacos
2              Chicken Bowl
7           Chicken Burrito
5      Chicken Crispy Tacos
22            Chicken Salad
13       Chicken Salad Bowl
6        Chicken Soft Tacos
28             Crispy Tacos
0                      Izze
1          Nantucket Nectar
34                    Salad
14               Steak Bowl
3             Steak Burrito
18       Steak Crispy Tacos
30              Steak Salad
24         Steak Salad Bowl
4          Steak Soft Tacos
17              Veggie Bowl
16           Veggie Burrito
35      Veggie Crispy Tacos
36             Veggi

## 9. Quantity and Price Relationships:

In [178]:
chipotle[['quantity', 'item_price']]

quantity  item_price
1            1        3.39
2            1        3.39
4            2       16.98
5            1       10.98
7            1       11.75
...        ...         ...
4617         1       11.75
4618         1       11.75
4619         1       11.25
4620         1        8.75
4621         1        8.75

[3335 rows x 2 columns]

In [180]:
chipotle.groupby('quantity').agg({'item_price': ['min', 'mean', 'median', 'max', 'count']})

item_price                               
                min       mean median    max count
quantity                                          
1              1.09   8.716340   8.99  11.89  3150
2              2.18  14.137407  17.50  23.78   162
3              3.75  18.277059  26.07  35.25    17
4              4.36  14.893333   5.00  35.00     6

**From above output we can analyze**:
1. The minimum price is 1.09 dollars per quantity of a item
2. The maximum price is 35.25 dollars per 3 quantity of a item
3. The maximum no. of orders is of single quantity which is total of 3150
4. The miinimum no. of orders is lowest when together ordering 4 quantities of a single item
5. There is no significant difference in average price for the different quantities

## 10. Data Integrity Check:

In [185]:
chipotle.reset_index(drop=True, inplace=True)
chipotle

order_id  quantity           item_name  \
0           1         1                Izze   
1           1         1    Nantucket Nectar   
2           2         2        Chicken Bowl   
3           3         1        Chicken Bowl   
4           4         1       Steak Burrito   
...       ...       ...                 ...   
3330     1833         1       Steak Burrito   
3331     1833         1       Steak Burrito   
3332     1834         1  Chicken Salad Bowl   
3333     1834         1  Chicken Salad Bowl   
3334     1834         1  Chicken Salad Bowl   

                                     choice_description  item_price  
0                                            Clementine        3.39  
1                                                 Apple        3.39  
2     Tomatillo-Red Chili Salsa (Hot), Black Beans, ...       16.98  
3     Fresh Tomato Salsa (Mild), Rice, Cheese, Sour ...       10.98  
4     Tomatillo Red Chili Salsa, Fajita Vegetables, ...       11.75  
...                                                 ...         ...  
3330  Fresh Tomato Salsa, Rice, Black Beans, Sour Cr...       11.75  
3331  Fresh Tomato Salsa, Rice, Sour Cream, Cheese, ...       11.75  
3332  Fresh Tomato Salsa, Fajita Vegetables, Pinto B...       11.25  
3333     Fresh Tomato Salsa, Fajita Vegetables, Lettuce        8.75  
3334  Fresh Tomato Salsa, Fajita Vegetables, Pinto B...        8.75  

[3335 rows x 5 columns]

## 11. Converting to CSV:

In [186]:
chipotle.to_csv('finalized_chipotle_dataset.csv')

## 12. Handling Categorical Data:

In [188]:
chipotle.item_name.unique()

array(['Izze', 'Nantucket Nectar', 'Chicken Bowl', 'Steak Burrito',
       'Steak Soft Tacos', 'Chicken Crispy Tacos', 'Chicken Soft Tacos',
       'Chicken Burrito', 'Canned Soda', 'Barbacoa Burrito',
       'Carnitas Burrito', 'Carnitas Bowl', 'Barbacoa Bowl',
       'Chicken Salad Bowl', 'Steak Bowl', 'Barbacoa Soft Tacos',
       'Veggie Burrito', 'Veggie Bowl', 'Steak Crispy Tacos',
       'Barbacoa Crispy Tacos', 'Veggie Salad Bowl',
       'Carnitas Soft Tacos', 'Chicken Salad', 'Canned Soft Drink',
       'Steak Salad Bowl', '6 Pack Soft Drink', 'Bowl', 'Burrito',
       'Crispy Tacos', 'Carnitas Crispy Tacos', 'Steak Salad',
       'Veggie Soft Tacos', 'Carnitas Salad Bowl', 'Barbacoa Salad Bowl',
       'Salad', 'Veggie Crispy Tacos', 'Veggie Salad', 'Carnitas Salad'],
      dtype=object)

In [190]:
item_map = {'6 Pack Soft Drink': 'Soft Drink Pack',
       'Barbacoa Bowl': 'Barbacoa',
       'Barbacoa Burrito': 'Barbacoa',
       'Bowl': 'Bowl',
       'Burrito': 'Burrito',
       'Canned Soda': 'Canned',
       'Canned Soft Drink': 'Canned',
       'Carnitas Bowl': 'Carnitas',
       'Carnitas Burrito': 'Carnitas',
       'Carnitas Crispy Tacos': 'Carnitas',
       'Carnitas Salad': 'Carnitas',
       'Carnitas Salad Bowl': 'Carnitas',
       'Carnitas Soft Tacos': 'Carnitas',
       'Chicken Bowl': 'Chicken',
       'Chicken Burrito': 'Chicken',
       'Chicken Crispy Tacos': 'Chicken',
       'Chicken Salad': 'Chicken',
       'Chicken Salad Bowl': 'Chicken',
       'Chicken Soft Tacos': 'Chicken',
       'Izze': 'Izze',
       'Nantucket Nectar': 'Nantucket',
       'Salad': 'Salad',
       'Steak Bowl': 'Steak',
       'Steak Burrito': 'Steak',
       'Steak Crispy Tacos': 'Steak',
       'Steak Salad': 'Steak',
       'Steak Salad Bowl': 'Steak',
       'Steak Soft Tacos': 'Steak',
       'Veggie Bowl': 'Veggie',
       'Veggie Burrito': 'Veggie',
       'Veggie Crispy Tacos': 'Veggie',
       'Veggie Salad': 'Veggie',
       'Veggie Salad Bowl': 'Veggie',
       'Veggie Soft Tacos': 'Veggie'}

In [194]:
chipotle = chipotle.assign(item_type = chipotle.item_name.map(item_map))
chipotle

## 13. Consistent Quantity and Price Units:

Already `quantity` is in int64 datatype and `item_price` is in float datatype (represented in dollars)

---
# END OF PROJECT